test_regress/t/t_vpi_multidim.cpp

```diff
@@ -0,0 +1,447 @@
+// -*- mode: C++; c-file-style: "cc-mode" -*-
+//*************************************************************************
+//
+// Copyright 2024 by Wilson Snyder. This program is free software; you can
+// redistribute it and/or modify it under the terms of either the GNU
+// Lesser General Public License Version 3 or the Perl Artistic License
+// Version 2.0.
+// SPDX-License-Identifier: LGPL-3.0-only OR Artistic-2.0
+//
+//*************************************************************************
+
+#ifdef IS_VPI
+
+#include "vpi_user.h"
+
+#include <cstdlib>
+
+#else
+
+#include "verilated.h"
+#include "verilated_vcd_c.h"
+#include "verilated_vpi.h"
+
+#include "Vt_vpi_multidim.h"
+#include "Vt_vpi_multidim__Dpi.h"
+#include "svdpi.h"
+
+#endif
+
+#include <cstdio>
+#include <cstring>
+#include <iostream>
+#include <random>
+#include <cstdlib>
+
+// These require the above. Comment prevents clang-format moving them
+#include "TestCheck.h"
+#include "TestSimulator.h"
+#include "TestVpi.h"
+
+#define FILENM "t_vpi_multidim.cpp"
+
+int errors = 0;
+
+// TEST START
+
+void _arr_iter_check(const char* name, int wordSize) {
+    TestVpiHandle arr_h = vpi_handle_by_name((PLI_BYTE8*)TestSimulator::rooted(name), NULL);
+    TEST_CHECK_NZ(arr_h);
+
+    {
+        // variable is size-4 RegArray
+        int vpitype = vpi_get(vpiType, arr_h);
+        TEST_CHECK_EQ(vpitype, vpiRegArray);
+        int vpisize = vpi_get(vpiSize, arr_h);
+        TEST_CHECK_EQ(vpisize, 4);
+    }
+
+    {
+        // can't iterate through RegArrays on a nested RegArray
+        TestVpiHandle arr_iter_h = vpi_iterate(vpiRegArray, arr_h);
+        TEST_CHECK_Z(vpi_scan(arr_iter_h));
+        arr_iter_h.freed();
+    }
+
+    if (!TestSimulator::is_questa()) {
+        // but we can access them by index (Questa can't)
+        for (int idx = 0; idx < 2; idx++) {
+            TestVpiHandle arr_elem_h = vpi_handle_by_index(arr_h, idx);
+            TEST_CHECK_NZ(arr_elem_h);
+            {
+                // first indexing yields size-2 RegArrays
+                int vpitype = vpi_get(vpiType, arr_elem_h);
+                TEST_CHECK_EQ(vpitype, vpiRegArray);
+                int vpisize = vpi_get(vpiSize, arr_elem_h);
+                TEST_CHECK_EQ(vpisize, 2);
```

Comment:
This pattern seems repeated a lot, make a subroutine?  Also since doing that check the low and high index are correct?